src/query/storage/m3/storage.go

```diff
@@ -257,20 +260,28 @@ func (s *m3storage) Write(
 	id := query.Tags.ID()
 	// TODO: Consider caching id -> identID
 	identID := ident.StringID(id)
-	common := &writeRequestCommon{
-		store:       s,
-		annotation:  query.Annotation,
-		unit:        query.Unit,
-		id:          identID,
-		tagIterator: storage.TagsToIdentTagIterator(query.Tags),
-		attributes:  query.Attributes,
-	}
+	tagIterator := storage.TagsToIdentTagIterator(query.Tags)
+
+	var (
+		wg       sync.WaitGroup
+		multiErr xerrors.MultiError
+	)
+
+	for _, datapoint := range query.Datapoints {
+		s.writeWorkerPool.Go(func() {
```

Comment:
I think you need to capture the datapoint variable here; surprised that there are no failing tests... might be good to add one that tries to write multiple datapoints and checks that the correct ones are being written?

Reply:
good point.